packages/timeline-state-resolver/src/service/stateHandler.ts

```diff
@@ -0,0 +1,122 @@
+import { Mappings, Timeline, TSRTimelineContent } from 'timeline-state-resolver-types'
+
+export class StateHandler<DeviceState, Command> {
+	private stateQueue: {
+		commands?: Command[]
+		state: Timeline.TimelineState<TSRTimelineContent>
+		deviceState: DeviceState
+		mappings: Mappings
+	}[] = []
+	private currentState:
+		| {
+				commands: Command[]
+				state: Timeline.TimelineState<TSRTimelineContent>
+				deviceState: DeviceState
+				mappings: Mappings
+		  }
+		| undefined = undefined
+
+	private clock: NodeJS.Timeout
+
+	constructor(
+		private convertTimelineStateToDeviceState: (
+			state: Timeline.TimelineState<TSRTimelineContent>,
+			mappings: Mappings
+		) => DeviceState,
+		private diffDeviceStates: (oldState: DeviceState, newState: DeviceState, mappings: Mappings) => Command[],
+		private executeCommand: (command: Command) => any
+	) {
+		this.clock = setInterval(() => {
+			// main clock to check if next state needs to be sent out
+			const nextState = this.stateQueue[0]
+
+			if (this.currentState && nextState && nextState.state.time <= Date.now()) this.executeNextStateChange()
```

Comment:
This looks like a floating promise to me. And because this is running on an interval, is there no risk of executing the next state change twice if for any weird reason [this part](https://github.com/nrkno/sofie-timeline-state-resolver/pull/236/files#diff-a8cbc2cdbe840e77a7479a09a30f6f41e81435fdc8e36b3562bb6a70b44f0ff5R97-R99) does not resolve within 20ms and the interval fires again?

Reply:
No, it shouldn't execute twice because it'll be removed from the state queue after the first execution. However, some other weirdness may ensue so I'll need to put in some logic to ensure we don't execute multiple `executeNextStateChange` at the same time